RecoEgamma/EgammaIsolationAlgos/src/EleTkIsolFromCands.cc

```diff
@@ -63,38 +62,90 @@ edm::ParameterSetDescription EleTkIsolFromCands::pSetDescript()
 std::pair<int,double> 
 EleTkIsolFromCands::calIsol(const reco::TrackBase& eleTrk,
 			    const pat::PackedCandidateCollection& cands,
-			    const edm::View<reco::GsfElectron>& eles)
+			    const PIDVeto pidVeto)const
 {
-  return calIsol(eleTrk.eta(),eleTrk.phi(),eleTrk.vz(),cands,eles);
+  return calIsol(eleTrk.eta(),eleTrk.phi(),eleTrk.vz(),cands,pidVeto);
 }
 
 std::pair<int,double> 
 EleTkIsolFromCands::calIsol(const double eleEta,const double elePhi,
 			    const double eleVZ,
 			    const pat::PackedCandidateCollection& cands,
-			    const edm::View<reco::GsfElectron>& eles)
+			    const PIDVeto pidVeto)const
 {
-
   double ptSum=0.;
   int nrTrks=0;
 
   const TrkCuts& cuts = std::abs(eleEta)<1.5 ? barrelCuts_ : endcapCuts_;
   
   for(auto& cand  : cands){
-    if(cand.charge()!=0){
-      const reco::Track& trk = cand.pseudoTrack(); 
-      double trkPt = std::abs(cand.pdgId())!=11 ? trk.pt() : getTrkPt(trk,eles);
-      if(passTrkSel(trk,trkPt,cuts,eleEta,elePhi,eleVZ)){	
-	ptSum+=trkPt;
+    if(cand.hasTrackDetails() && cand.charge()!=0 && passPIDVeto(cand.pdgId(),pidVeto)){
+      const reco::Track& trk = cand.pseudoTrack();
+      if(passTrkSel(trk,trk.pt(),cuts,eleEta,elePhi,eleVZ)){	
+	ptSum+=trk.pt();
 	nrTrks++;
       }
     }
   }
   return {nrTrks,ptSum};	
 }
-	
 
-bool EleTkIsolFromCands::passTrkSel(const reco::Track& trk,
+
+
+std::pair<int,double> 
+EleTkIsolFromCands::calIsol(const reco::TrackBase& eleTrk,
+			    const reco::TrackCollection& tracks)const
+{
+  return calIsol(eleTrk.eta(),eleTrk.phi(),eleTrk.vz(),tracks);
+}
+
+std::pair<int,double> 
+EleTkIsolFromCands::calIsol(const double eleEta,const double elePhi,
+			    const double eleVZ,
+			    const reco::TrackCollection& tracks)const
+{
+  double ptSum=0.;
+  int nrTrks=0;
+
+  const TrkCuts& cuts = std::abs(eleEta)<1.5 ? barrelCuts_ : endcapCuts_;
+  
+  for(auto& trk  : tracks){
+    if(passTrkSel(trk,trk.pt(),cuts,eleEta,elePhi,eleVZ)){	
+      ptSum+=trk.pt();
+      nrTrks++;
+    }
+  }
+  return {nrTrks,ptSum};	
+}	
+
+bool EleTkIsolFromCands::passPIDVeto(const int pdgId,const EleTkIsolFromCands::PIDVeto veto)
+{
+  int pidAbs = std::abs(pdgId);
+  switch (veto){
+  case PIDVeto::NONE:
+    return true;
+  case PIDVeto::ELES:
+    if(pidAbs==11) return false;
+    else return true;
+  case PIDVeto::NONELES:
+    if(pidAbs==11) return true;
+    else return false;
+  default:
+    throw cms::Exception("CodeError") <<"invalid PIDVeto "<<static_cast<int>(veto)<<", EleTkIsolFromCands::PIDVeto has been updated without EleTkIsolFromCands::passPIDVeto also being updated";
+  }
+}
+
+EleTkIsolFromCands::PIDVeto EleTkIsolFromCands::pidVetoFromStr(const std::string& vetoStr) 
+{
+  if(vetoStr=="none") return PIDVeto::NONE;
+  else if(vetoStr=="eles") return PIDVeto::ELES;
+  else if(vetoStr=="noneles") return PIDVeto::NONELES;
```

Comment:
is there a reason the lower case is used  for string compared to upper case for C++?

Reply:
yes, not a good one though. Its my private naming convention (I put enums caps to distinguish them although that is less necessary now with proper scoped enums and strings are lowercase). I'm perfectly willing to change them to something else, what would be the reco groups recommendation? "kEles, kNone, kNonEles" would probably be the best fit?

Reply:
I wasn't going after ```NONE ``` by itself.
My comment was about the difference between ```none``` as a string and ```NONE``` as a C++ , aiming to reduce the number of variants one needs to remember to do translation.